appium.js

```diff
@@ -138,31 +158,76 @@ function connectCreateNewSession () {
       https,
     });
 
+    // Try initializing it. If it fails, kill it and send error message to sender
     try {
       let p = driver.init(desiredCapabilities);
       event.sender.send('appium-new-session-successful');
       await p;
       event.sender.send('appium-new-session-ready');
-
     } catch (e) {
-
       // If the session failed, delete it from the cache
-      delete sessionDrivers[event.sender.id];
-      event.sender.send('appium-new-session-failed', JSON.parse(e.data).value.message);
+      await killSession(event.sender);
+      event.sender.send('appium-new-session-failed');
     }
 
   });
 }
 
+/**
+ * When a Session Window makes method request, find it's corresponding driver, execute requested method
+ * and send back the result
+ */
+function connectClientMethodListener () {
+  ipcMain.on('appium-client-command-request', async (evt, data) => {
+    const {methodName, args = [], xpath} = data;
+    let driver = sessionDrivers[evt.sender.id];
+
+    let source, screenshot;
+    try {
+      if (methodName === 'quit') {
+        evt.sender.send('appium-session-done');
+        await killSession(evt.sender);
+      } else {
+
+        // Execute the requested method
+        if (methodName !== 'source') {
+          if (xpath) {
+            await driver.elementByXPath(xpath)[methodName](...args);
+          } else {
+            await driver[methodName](...args);
+          }
+        }
+        await Bluebird.delay(500);
```

Comment:
should probably add a comment as to why we're doing this (to allow any interaction to happen before we get the new source)